### PR TITLE
Refactor 'showCancelMediaUploadDialog' to Prevent Dismissal on Screen Rotation

### PR DESCRIPTION
**Pull request by SiobhyB**
## Description

Partial fix for https://github.com/wordpress-mobile/WordPress-Android/issues/14676

The `showCancelMediaUploadDialog`, which is shown when tapping an image that's still in the process of uploading, is currently dismissed/lost with a screen rotation. 

This can lead to buggy behavior, as shown in the following screenshots, where the image will permanently look like it's still uploading, with any attempts to tap on it resulting in the dialog appearing again, even though the image has been uploaded to the site's media library:

| Before Rotation  | After Rotation |
| ------------- | ------------- |
|  |  |

This PR seeks to fix the above issue by refactoring the dialog so that it survives screen rotation. A custom dialog fragment, `GutenbergDialogFragment`, is used to achieve this.

### GutenbergDialogFragment

A [new `GutenbergDialogFragment` class](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergDialogFragment.kt) was introduced in https://github.com/wordpress-mobile/WordPress-Android/pull/14503. This class is based on [the `BasicFragmentDialog` class](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/BasicFragmentDialog.kt), with some key differences being the removal of the third `setNeutralButton`, [an addition of a mediaId](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-8889a43bfa80f8522889ae2f69695ac5eda5b29189bb29c9915a9c4825dae6e7R39) to the constructor, and the ability for [the parent fragment to implement the positive/negative button interfaces](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-8889a43bfa80f8522889ae2f69695ac5eda5b29189bb29c9915a9c4825dae6e7R86). 

Part of the goal behind creating the new `GutenbergDialogFragment` class was to allow for easier refactoring of the other dialogs listed in https://github.com/wordpress-mobile/WordPress-Android/issues/14676, which are all currently lost/dismissed with a screen rotation. With that goal in mind, [the mediaId param is refactored with this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/14810/commits/572e9b3b5ff204edf2d2800466e2bc579c30b02b) to utilize generics and accept any data type. Its [message param has also been made optional](https://github.com/wordpress-mobile/WordPress-Android/pull/14810/commits/b0cefc8938f514455882588406edef11b5c1c72b), as not all dialogs declared in `GutenbergEditorFragment` include a message. 

The aforementioned changes make it relatively straightforward to [refactor `showCancelMediaUploadDialog` to use the custom dialog fragment](https://github.com/wordpress-mobile/WordPress-Android/pull/14810/commits/84a7b12d48dbfad9bcebe3ca6c394d362d952721).

## Testing Steps

* Navigate to the post editor.
* Add a new image block and choose the option to upload an image directly from your device.
* Immediately tap on the image block as the upload begins in order to generate the "cancel" dialog.
* Switch the orientation of the device and confirm that the dialog is still in place.
* Tap the "no" or "yes" buttons after rotating the screen to confirm that they still work as expected.
* Try a few combinations of rotations and button pressing to confirm there are no unintended or buggy side effects to this recent change.

## Regression Notes
1. Potential unintended areas of impact

The main unintended side effect of this PR would be for the dialog's existing "yes" or "no" flows to break.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I following the outlined testing steps, with various combinations of rotations and button pressing to see if I could create any problems or issues.

3. What automated tests I added (or what prevented me from doing so)

Automated tests didn't feel warranted given the size of this change.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by SiobhyB** [comment]
👋 @planarvoid, as [we previously talked a bit](https://github.com/wordpress-mobile/WordPress-Android/issues/14676#issuecomment-843081165) about this issue, I wondered if you might have some availability this week to review my first pass at fixing one of the dialogs? Let me know. :)

**Comment by SiobhyB** [comment]
@planarvoid, I requested you as a reviewer, but let me know if you won't be able to get to this week or next, and I can find someone else. :)

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergDialogFragment.kt [hunk 1]
```diff
@@ -7,19 +7,20 @@ import android.os.Bundle
 import androidx.appcompat.app.AppCompatDialogFragment
 import androidx.fragment.app.Fragment
 import com.google.android.material.dialog.MaterialAlertDialogBuilder
+import java.io.Serializable
 
-class GutenbergDialogFragment() : AppCompatDialogFragment() {
+class GutenbergDialogFragment<T : Serializable?>() : AppCompatDialogFragment() {
     private lateinit var mTag: String
-    private lateinit var mMessage: CharSequence
+    private var mMessage: CharSequence? = null
```
Comment (planarvoid) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergDialogFragment.kt:14]: we shouldn't use the Hungarian notation in Kotlin 👍

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergDialogFragment.kt [hunk 2]
```diff
@@ -7,19 +7,20 @@ import android.os.Bundle
...
     private var mNegativeButtonLabel: CharSequence? = null
-    private var mId: Int = 0
+    private var mDataFromGutenberg: T? = null
     private var dismissedByPositiveButton = false
     private var dismissedByNegativeButton = false
 
-    interface GutenbergDialogPositiveClickInterface {
-        fun onGutenbergDialogPositiveClicked(instanceTag: String, id: Int)
+    interface GutenbergDialogPositiveClickInterface<T> {
+        fun onGutenbergDialogPositiveClicked(instanceTag: String, dataFromGutenberg: Any?)
```
Comment (planarvoid) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergDialogFragment.kt:23]: I assume the type of `dataFromGutenberg` should be `T`, not `Any`

**Comment by planarvoid** [comment]
Thanks for the PR! 
I don't think it's a good idea to have the `GutenbergDialogFragment` as generic (I'm honestly not sure how that would work with creating and restoring fragments in the fragment manager). What kind of data do you expect to pass around? Could we maybe use a Parcelable? or something stored in a database? Or an array? I think it's usually cleaner to not do inheritance in cases like this. Doing an implementation of the dialog fragment for each case should be simple enough, right? 
Also I think it would be much cleaner to do something like a `GutenbergDialogViewModel` - a class that would be responsible for showing, hiding and handling the dialogs, instead of using the old approach with the fragment implementing the positive/negative interface. This would be much more in line with the way we want the app to be written (check `StatsSiteSelectionViewModel` and `StatsWidgetSiteSelectionDialogFragment`).

**Comment by SiobhyB** [comment]
> I don't think it's a good idea to have the GutenbergDialogFragment as generic (I'm honestly not sure how that would work with creating and restoring fragments in the fragment manager).

Thank you for your candid feedback, would you be able to help me understand the concern with how the fragment manager creates/restores fragments using this approach? What type of problems could it cause?

> What kind of data do you expect to pass around? Could we maybe use a Parcelable? or something stored in a database? Or an array? I think it's usually cleaner to not do inheritance in cases like this.

The data varies for each of the six dialogs in `GutenbergEditorFragment` that need to be refactored. In the case of this PR, the data is a media ID (an integer), but there are other dialogs that pass media files (stored in an ArrayList).

I can definitely research the other options you mentioned and will seek to gain a better understanding of why they're more suitable as part of that research.

> Doing an implementation of the dialog fragment for each case should be simple enough, right?

To make sure I understand you correctly, are you suggesting that I create six separate dialog fragments for each of the dialogs declared in `GutenbergEditorFragment`? That does seem simple on the face of it, but also like it'll mean a lot of copying/pasting and duplicated code. I'll need to take some time to research and understand the concerns you've raised in order to understand the reason why this is a better approach.

> Also I think it would be much cleaner to do something like a GutenbergDialogViewModel - a class that would be responsible for showing, hiding and handling the dialogs, instead of using the old approach with the fragment implementing the positive/negative interface. This would be much more in line with the way we want the app to be written (check StatsSiteSelectionViewModel and StatsWidgetSiteSelectionDialogFragment).

Thank you for the context around the pattern I followed being an older approach. It's definitely tricky to distinguish between old and new approaches when looking for patterns as a junior exploring the codebase. In this case, the pattern I copied was from [the BasicFragmentDialog class](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/BasicFragmentDialog.kt). Do you think it'd be worthwhile for me to create a PR to add a summary of what you mentioned around the preferable approach to the comments of that file?

Thanks again in advance for your patience with my questions as I learn and sharing your knowledge here. 🙇‍♀️

**Comment by planarvoid** [comment]
thanks for the reply! 
> Thank you for your candid feedback, would you be able to help me understand the concern with how the fragment manager creates/restores fragments using this approach? What type of problems could it cause?
Sorry if my feedback sounded harsh. It wasn't my intention. The generics get stripped during compilation so if I understand it correctly, something like this: 
```
FragmentA<Integer> fragment = (FragmentA<Integer>) fm.findFragmentByTag(TAG);
```
is not a safe cast. This is what Android Studio tells you: 
``` 
Unchecked cast: 'androidx.fragment.app.Fragment' to 'org.wordpress.android.ui.mlp.ModalLayoutPickerFragment<java.lang.Integer>'
```
I think it will still work but I wouldn't do it. A check like this: 
```
fragment is FragmentA<Integer>
```
will not work in Kotlin. 

> I can definitely research the other options you mentioned and will seek to gain a better understanding of why they're more suitable as part of that research.

I think `Parcelable` might work well in this case. 

> To make sure I understand you correctly, are you suggesting that I create six separate dialog fragments for each of the dialogs declared in GutenbergEditorFragment? That does seem simple on the face of it, but also like it'll mean a lot of copying/pasting and duplicated code. I'll need to take some time to research and understand the concerns you've raised in order to understand the reason why this is a better approach.

What I meant is that you can have 6 children without any logic that would just inherit from the `GutenbergEditorFragment` and define their own resources/type and pass it up to the parent using the constructor. It could improve readibility. 

> Thank you for the context around the pattern I followed being an older approach. It's definitely tricky to distinguish between old and new approaches when looking for patterns as a junior exploring the codebase. In this case, the pattern I copied was from the BasicFragmentDialog class. Do you think it'd be worthwhile for me to create a PR to add a summary of what you mentioned around the preferable approach to the comments of that file?

I think we can add that. That's a good idea! Sadly we don't have enough time to deal with the technical debt so we keep the old files around as long as they are used somewhere.

**Comment by SiobhyB** [comment]
I'm going to go ahead to close this PR as it's not currently being worked on or a priority.

**Pull request closed without merging**